### PR TITLE
[FrameworkBundle][HttpKernel] Display warmers duration on debug verbosity for `cache:clear` command

### LINKED ISSUE #48994: [DX][Container] When doing cache:clear with warmup and (very) verbose / debug output, display which warmers are running
### Description

When running

```
bin/console cache:clear -vvv
```

the output is typically

```
 // Clearing the cache for the dev environment with debug true                                                          

 // Warming up cache...      
 
 // Removing old build and cache directory...                                                                           

 // Finished                                                                                                            
                                                                                                                        
 [OK] Cache for the "dev" environment (debug=true) was successfully cleared.                                            
                                                                                                                        
```
                                                                                                                        
You get the same level of verbosity as you do with just `-v`.

What could happen here is we could print:
1. which warmers are running
2. how much time each one took

This would allow to demistify what is the clear cache doing and what's taking so long.

### Example

```
bin/console cache:clear -vvv

 // Warming up cache...      
 
^ "Symfony\Bundle\FrameworkBundle\CacheWarmer\CachePoolClearerCacheWarmer"  - 230ms
^ "Symfony\Bundle\FrameworkBundle\CacheWarmer\ConfigBuilderCacheWarmer" - 13ms
^ "Symfony\Bundle\FrameworkBundle\CacheWarmer\TranslationsCacheWarmer" - 23ms
```

etc.

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #48994
| License       | MIT
| Doc PR        | Todo

From the idea of the issue it fixes. When dealing with huge code base, the cache/warmup process can be a bit long. It would be convenient for developers to understand where it does take time during the process. Here is the result of `cache:clear` ran with debug verbosity now:

image

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php [hunk 1]
```diff
@@ -24,6 +24,7 @@
 use Symfony\Component\Filesystem\Filesystem;
 use Symfony\Component\Finder\Finder;
 use Symfony\Component\HttpKernel\CacheClearer\CacheClearerInterface;
+use Symfony\Component\HttpKernel\Event\Cache\CacheWarmerAggregateItemFinishedEvent;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php]: Can we achieve this feature without introducing any events? I don't think we want to expose a new hook point for this.

File: src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php [hunk 2]
```diff
@@ -37,14 +38,15 @@
...
 {
-    private CacheClearerInterface $cacheClearer;
     private Filesystem $filesystem;
+    private EventDispatcher $eventDispatcher;
 
-    public function __construct(CacheClearerInterface $cacheClearer, Filesystem $filesystem = null)
-    {
+    public function __construct(
+        private readonly CacheClearerInterface $cacheClearer,
+        Filesystem $filesystem = null
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php]: please revert all unrelated CS changes from this PR (CPP)

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php [hunk 3]
```diff
@@ -20,20 +23,17 @@
...
     private bool $onlyOptionalsEnabled = false;
 
     /**
      * @param iterable<mixed, CacheWarmerInterface> $warmers
      */
-    public function __construct(iterable $warmers = [], bool $debug = false, string $deprecationLogsFilepath = null)
-    {
-        $this->warmers = $warmers;
-        $this->debug = $debug;
-        $this->deprecationLogsFilepath = $deprecationLogsFilepath;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: revert CPP

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php [hunk 4]
```diff
@@ -46,7 +46,7 @@ public function enableOnlyOptionalWarmers(): void
         $this->onlyOptionalsEnabled = $this->optionalsEnabled = true;
     }
 
-    public function warmUp(string $cacheDir): array
+    public function warmUp(string $cacheDir, EventDispatcherInterface $eventDispatcher = null): array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: let's accept $io instead and use it to print some message when passed

**Comment by alexandre-daubois** [comment]
That's all updated! Thanks for suggesting this more pragmatic way 🙏

**Comment by alexandre-daubois** [comment]
That's all updated! Thanks for suggesting this more pragmatic way 🙏

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php [hunk 5]
```diff
@@ -46,7 +48,7 @@ public function enableOnlyOptionalWarmers(): void
         $this->onlyOptionalsEnabled = $this->optionalsEnabled = true;
     }
 
-    public function warmUp(string $cacheDir): array
+    public function warmUp(string $cacheDir, SymfonyStyle $io = null): array
```
Comment (alexislefebvre) [inline src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php:51]: Is it a BC break to add a required argument here?

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php [hunk 6]
```diff
@@ -93,12 +95,17 @@ public function warmUp(string $cacheDir): array
...
+                $start = microtime(true);
                 foreach ((array) $warmer->warmUp($cacheDir) as $item) {
                     if (is_dir($item) || (str_starts_with($item, \dirname($cacheDir)) && !is_file($item))) {
                         throw new \LogicException(sprintf('"%s::warmUp()" should return a list of files or classes but "%s" is none of them.', $warmer::class, $item));
                     }
                     $preload[] = $item;
                 }
+
+                if ($io?->isDebug()) {
+                    $io->info(sprintf('"%s" finished warmup in %.2f ms.', $warmer::class, (microtime(true) - $start) * 1000.0));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: `$io->info(sprintf('"%s" completed in %.2fms.', $warmer::class, 1000 * (microtime(true) - $start)));`

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: `%0.2f` actually?

Comment (alexandre-daubois) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: Suits me fine, that's updated!

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php [hunk 7]
```diff
@@ -93,12 +95,17 @@ public function warmUp(string $cacheDir): array
...
+                $start = microtime(true);
                 foreach ((array) $warmer->warmUp($cacheDir) as $item) {
                     if (is_dir($item) || (str_starts_with($item, \dirname($cacheDir)) && !is_file($item))) {
                         throw new \LogicException(sprintf('"%s::warmUp()" should return a list of files or classes but "%s" is none of them.', $warmer::class, $item));
                     }
                     $preload[] = $item;
                 }
+
+                if ($io?->isDebug()) {
+                    $io->info(sprintf('"%s" completed in %0.2f ms.', $warmer::class, 1000 * (microtime(true) - $start)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: I'd remove the space before ms (as in my previous suggestion ;) )

Comment (alexandre-daubois) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: Oops sorry, missed it 👀 That's updated as well!

Comment (alexisLefebvre) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: Is there some alternative syntax to write it?

```php
$io->info(sprintf('"%s" completed in %0.2fms.'
```

It's hard to read `%0.2fms` and figure out that `%0.2f` is a placeholder.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: What's important is what is displayed to end users, and this space is not the best we can do :)

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php (hunk 7 above)
Comment (alexisLefebvre) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php]: Of course, I was thinking from a maintainer's point of view.

I couldn't find a better way to write this so it looks like we have no other choice here.

File: src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php [hunk 8]
```diff
@@ -93,12 +95,17 @@ public function warmUp(string $cacheDir): array
...
+                $start = microtime(true);
                 foreach ((array) $warmer->warmUp($cacheDir) as $item) {
                     if (is_dir($item) || (str_starts_with($item, \dirname($cacheDir)) && !is_file($item))) {
                         throw new \LogicException(sprintf('"%s::warmUp()" should return a list of files or classes but "%s" is none of them.', $warmer::class, $item));
                     }
                     $preload[] = $item;
                 }
+
+                if ($io?->isDebug()) {
+                    $io->info(sprintf('"%s" completed in %0.2fms.', $warmer::class, 1000 * (microtime(true) - $start)));
```
Comment (alexisLefebvre) [inline src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php:107]:
```suggestion
                    $io->info(strtr('"%class%" completed in %time%ms.', [
                        '%class%' => $warmer::class,
                        '%time%' => sprintf('%0.2f', 1000 * (microtime(true) - $start)),
                    ]));
```

This syntax is easier to read, but ultimately it may not be worth it.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php:107]: not worth it yes :) thanks for having a look

**Comment by nicolas-grekas** [comment]
Thank you @alexandre-daubois.

**Comment by dkarlovi** [comment]
Thanks for working on this @alexandre-daubois! \o/
